_posts/2016-08-10-dynamic-loading.md

```diff
@@ -0,0 +1,91 @@
+---
+title: Dynamic Code Loading on a MCU
+authors:
+  - brghena
+  - alevy
+---
+
+One key feature of Tock is the ability to load and run multiple applications
+simultaneously. In a modern computer, the OS uses the Memory Management Unit
+(MMU) to provide a virtual address space for each process. The code for each
+process is written assuming that it is the only code in the world and that it
+can place memory at any address it pleases. The MMU handles translating these
+virtual addresses into a physical address in real memory, which is shared
+between all processes. Unfortunately, in the world of embedded systems MMUs are
+not available. Processors like the ARM Cortex-M series omit them because they
+aren't necessary for traditional single-process embedded devices.
+
+A common approach to handle this issue is to assign addresses to applications
+at compile-time. For example, Application A can be placed at address 0x21000
+and Application B can be placed at address 0x22000. This runs into problems
+when Application A grows in size in the future. Moreover, with Tock we want to
+be able to handle dynamically adding, updating, and removing applications at
+run time. Assigning each an address in advance simply isn't possible.
```

Comment:
Actually, protection is no different than what we are doing now. You could have address slots for each application and could use the MPU to protect each slot.